control-plane/agents/src/bin/core/controller/reconciler/volume/garbage_collector.rs

```diff
@@ -250,6 +258,53 @@ async fn disown_unused_replicas(
     GarbageCollector::squash_results(results)
 }
 
+/// Volume replicas could have been possibly expanded as part of volume resize,
+/// but the volume expansion failed overall leaving behind expanded replicas. So we
+/// bring back the replica size to what is expected by the volume spec.
+#[tracing::instrument(level = "debug", skip(context, volume), fields(volume.uuid = %volume.uuid(), request.reconcile = true))]
+async fn revert_expanded_replicas(
+    volume: &mut OperationGuardArc<VolumeSpec>,
+    context: &PollContext,
+) -> PollResult {
+    let target = context.specs().volume_target_nexus_rsc(volume.as_ref());
+    if let Some(tgt) = target {
+        let nexus_size = tgt.lock().size;
+        debug_assert_eq!(nexus_size, volume.as_ref().size);
+        if nexus_size != volume.as_ref().size {
+            // This should never happen, but just ensuring we don't touch replicas of such a volume.
+            volume.warn_span(|| tracing::info!(volume_size = %volume.as_ref().size, nexus_size = %nexus_size, "Nexus and Volume size mismatch!"));
+            return PollResult::Ok(PollerState::Idle);
+        }
+    }
+
+    for replica in context.specs().volume_replicas(volume.uuid()) {
+        let mut replica = match replica.operation_guard() {
+            Ok(guard) => guard,
+            Err(_) => continue,
+        };
+
+        if replica.as_ref().size != volume.as_ref().size {
+            let replica_state = context.registry().replica(replica.uuid()).await?;
+            volume.info_span(|| tracing::info!(replica = %replica.as_ref().uuid, replica_size = %replica.as_ref().size, "Reclaiming space from replica. Volume size {}", volume.as_ref().size));
+            // Resize replica back to volume spec size.
+            replica
+                .resize(
+                    context.registry(),
+                    &ResizeReplica::new(
+                        &replica_state.node,
+                        replica.as_ref().pool_name(),
+                        None,
+                        replica.uuid(),
+                        volume.as_ref().size,
+                    ),
+                )
+                .await?;
```

Comment:
hmm can this be done with ongoing IO with no disruption?

Reply:
good point. There won't be any disruption to IO. However, the second part of resize lvol call where it updates bdev block count(after adjusting clusters) will keep running into EBUSY till bdev has open descriptors.
Wondering if that can troublesome much.

Reply:
Simplest would be skip doing this if we find target above at line 270

Reply:
As discussed, not attempting reclaim in case a volume target is found.